File: notes/examples/rectangle/Rectangle.cpp
#include <Rcpp.h>
using namespace Rcpp;

class Rectangle {
public:
  Rectangle(double width, double height):
    width_(width), height_(height) { }
  double area() { return width_ * height_; }
private:
  double width_, height_;
};

RCPP_MODULE(Rectangle_Module) {
  class_<Rectangle>("Rectangle")
  
  .constructor<double, double>()
  .method("area", &Rectangle::area)
  ;
}

File: notes/examples/rectangle/rectangle.R

Rcpp::sourceCpp("rectangle.cpp")

r <- new(Rectangle, 2.5, 3)
r$area()

### S4 dispatch ###

# helper function to create objects
Rectangle <- function(width, height) {
  if(min(width,height) < 0) stop("width and height must be nonnegative.")
  new("Rcpp_Rectangle", width, height)
}

# generic function and method for area
setGeneric("area", function(object) standardGeneric("area"))
setMethod("area", "Rcpp_Rectangle", function(object) object$area())

Rectangle(-1,1)
rec <- Rectangle(2.5, 3)
area(rec)
